src/esp/gfx/PbrShader.cpp

```diff
@@ -0,0 +1,484 @@
+// LICENSE file in the root directory of this source tree.
+
+#include "PbrShader.h"
+
+#include <Corrade/Containers/Reference.h>
+#include <Corrade/Utility/Assert.h>
+#include <Corrade/Utility/Debug.h>
+#include <Corrade/Utility/DebugStl.h>
+#include <Corrade/Utility/FormatStl.h>
+#include <Corrade/Utility/Resource.h>
+#include <Magnum/GL/Context.h>
+#include <Magnum/GL/Extensions.h>
+#include <Magnum/GL/Shader.h>
+#include <Magnum/GL/Texture.h>
+#include <Magnum/GL/Version.h>
+#include <Magnum/ImageView.h>
+#include <Magnum/Math/Color.h>
+#include <Magnum/Math/Matrix3.h>
+#include <Magnum/Math/Matrix4.h>
+#include <Magnum/PixelFormat.h>
+
+#include "esp/core/esp.h"
+#include "esp/io/io.h"
+
+#include <sstream>
+
+// This is to import the "resources" at runtime. When the resource is
+// compiled into static library, it must be explicitly initialized via this
+// macro, and should be called *outside* of any namespace.
+static void importShaderResources() {
+  CORRADE_RESOURCE_INITIALIZE(ShaderResources)
+}
+
+namespace Mn = Magnum;
+namespace Cr = Corrade;
+
+namespace esp {
+namespace gfx {
+
+namespace {
+enum TextureUnit : uint8_t {
+  BaseColor = 0,
+  Roughness = 1,
+  Metallic = 2,
+  Normal = 3,
+  // NoneRoughnessMetallic or OcclusionRoughnessMetallic texture
+  Combined = 4,
+};
+}  // namespace
+
+PbrShader::Flags PbrShader::generateCorrectFlags(Flags originalFlags) {
+  Flags result = originalFlags;
+  // NOTE:
+  // The priority of different kind of textures is as follows (priority
+  // means if two textures with different priorities exist at the same time,
+  // shader will adopt the texture with the higher priority, and ignore the
+  // other one.)
+  // 0 (highest): OcclusionRoughnessMetallicTexture
+  // 1          : NoneRoughnessMetallicTexture
+  // 2          : RoughnessTexture, MetallicTexture
+  if (result & Flag::OcclusionRoughnessMetallicTexture) {
+    result &= ~Flag::NoneRoughnessMetallicTexture;
+    result &= ~Flag::RoughnessTexture;
+    result &= ~Flag::MetallicTexture;
+  } else if (result & Flag::NoneRoughnessMetallicTexture) {
+    result &= ~Flag::RoughnessTexture;
+    result &= ~Flag::MetallicTexture;
+  }
+
+  return result;
+}
+
+PbrShader::PbrShader(Flags originalFlags, unsigned int lightCount)
+    : flags_(generateCorrectFlags(originalFlags)), lightCount_(lightCount) {
+  if (!Cr::Utility::Resource::hasGroup("default-shaders")) {
+    importShaderResources();
+  }
+
+#ifdef MAGNUM_TARGET_WEBGL
+  Mn::GL::Version glVersion = Mn::GL::Version::GLES300;
+#else
+  Mn::GL::Version glVersion = Mn::GL::Version::GL330;
+#endif
+
+  // this is not the file name, but the group name in the config file
+  // see Shaders.conf in the shaders folder
+  const Cr::Utility::Resource rs{"default-shaders"};
+
+  Mn::GL::Shader vert{glVersion, Mn::GL::Shader::Type::Vertex};
+  Mn::GL::Shader frag{glVersion, Mn::GL::Shader::Type::Fragment};
+
+  std::stringstream attributeLocationsStream;
```

Comment:
@morsa: I do not use the `generic.glsl`. I used another way as I wrote here. I think it is better.
In your case, you will have to sync the locations between `generic.glsl` and the `Generic3D` if I understand you correctly. I do not have this issue, no matter how you refactor and change the locations in the Generic3D in the future.
Thoughts?

Reply:
Yep, that's what [the documentation suggests to do](https://doc.magnum.graphics/magnum/structMagnum_1_1Shaders_1_1Generic.html#Shaders-Generic-custom) as well. The `generic.glsl` is an implementation detail in Magnum, not meant to be used outside.

> you will have to sync the locations between

Yep, but I [have that covered by tests](https://github.com/mosra/magnum/blob/ba156d329a4347f2c58a2d63d71a2db645666ce9/src/Magnum/Shaders/Test/GenericTest.cpp#L54-L83). I chose this approach for the library because that means I don't need to do heavy string processing on every shader instantiation. It matters for certain use cases.

Reply:
I see. Thanks.